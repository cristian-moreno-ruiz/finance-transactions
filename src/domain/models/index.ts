export { TransactionModel } from 'src/domain/models/transaction-model';
export { TransactionStatus } from 'src/domain/models/transaction-status';
