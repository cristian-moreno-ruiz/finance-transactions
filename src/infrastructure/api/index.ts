export { TransactionRouter } from 'src/infrastructure/api/transaction-router';
export { TransactionController } from './transaction-controller';
