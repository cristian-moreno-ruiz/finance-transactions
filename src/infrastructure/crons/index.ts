export { TransactionIngestionCron } from 'src/infrastructure/crons/transaction-ingestion-cron';
