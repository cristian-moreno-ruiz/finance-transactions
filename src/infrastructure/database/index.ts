export { mongoInit } from 'src/infrastructure/database/mongo-init';
