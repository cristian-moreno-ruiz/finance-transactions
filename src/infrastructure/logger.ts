export class Logger {
	public static log(message?: any, ...optionalParams: any[]): void {
		console.log(message, optionalParams);
	}
	public static error(message?: any, ...optionalParams: any[]): void {
		console.error(message, optionalParams);
	}
}
