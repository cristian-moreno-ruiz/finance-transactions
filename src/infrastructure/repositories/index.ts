export { MongoTransactionRepository } from 'src/infrastructure/repositories/mongo-transaction-repository';
export { TransactionSchema, MongoTransactionModel } from 'src/infrastructure/repositories/mongo-transaction-schema';
