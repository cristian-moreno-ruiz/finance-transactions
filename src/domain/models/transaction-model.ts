import { TransactionStatus } from 'src/domain/models';
import { FinanceSourceEnum } from 'src/infrastructure/sources';

export interface TransactionModel {
	id: string;
	externalTransactionId: number;
	externalAccountId: number;
	title: string;
	description: string;
	date: Date;
	amount: number;
	currency: string;
	status: TransactionStatus;
	source: FinanceSourceEnum;
}
