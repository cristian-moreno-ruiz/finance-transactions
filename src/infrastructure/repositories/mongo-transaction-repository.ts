import { TransactionModel } from 'src/domain/models';
import { TransactionStatus } from 'src/domain/models/transaction-status';
import { TransactionRepositoryPort } from 'src/domain/ports';
import { MongoTransactionModel } from 'src/infrastructure/repositories';

export class MongoTransactionRepository implements TransactionRepositoryPort {
	public async getTransactionsInRange(
		from: Date,
		to: Date,
		status?: TransactionStatus,
	): Promise<TransactionModel[]> {
		const query: Record<string, unknown> = { date: { $gte: from, $lte: to } };
		if (status) {
			query.status = status;
		}

		const transactions = await MongoTransactionModel.find(query).lean().exec();
		return transactions;
	}

	public async refreshTransactions(transactions: TransactionModel[]): Promise<boolean> {
		const bulkOps = transactions.map((transaction) => ({
			updateOne: {
				filter: {
					externalTransactionId: transaction.externalTransactionId,
					externalAccountId: transaction.externalAccountId,
					source: transaction.source,
				},
				update: { ...transaction },
				upsert: true, // Insert if not found
			},
		}));

		const result = await MongoTransactionModel.bulkWrite(bulkOps);
		return !!result.ok;
	}
}
