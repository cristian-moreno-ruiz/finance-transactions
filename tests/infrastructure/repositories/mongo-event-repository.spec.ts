import { TransactionStatus } from 'src/domain/models/transaction-status';
import { MongoTransactionModel, MongoTransactionRepository } from 'src/infrastructure/repositories';
import { multipleTransactionsFixture } from 'tests/__fixtures__/transaction-model.fixtures';
import { createMockMongoTransactionModel } from 'tests/__mocks__/mongo-transaction-model.mock';
jest.mock('src/infrastructure/repositories/mongo-transaction-schema');

describe('MongoTransactionRepository', () => {
	let repository: MongoTransactionRepository;
	let mockedMongoTransactionModel: jest.Mocked<typeof MongoTransactionModel>;

	beforeEach(() => {
		mockedMongoTransactionModel = createMockMongoTransactionModel();
		repository = new MongoTransactionRepository();
	});

	afterEach(() => {
		jest.clearAllMocks();
	});

	describe('getTransactionsInRange', () => {
		it('should query transactions within a given date range', async () => {
			await repository.getTransactionsInRange(new Date('2023-01-01'), new Date('2023-01-03'));

			expect(mockedMongoTransactionModel.find).toHaveBeenCalledWith({
				date: { $gte: new Date('2023-01-01'), $lte: new Date('2023-01-03') },
			});
		});

		it('should filter by status if provided', async () => {
			await repository.getTransactionsInRange(
				new Date('2023-01-01'),
				new Date('2023-01-03'),
				TransactionStatus.COMPLETED,
			);

			expect(mockedMongoTransactionModel.find).toHaveBeenCalledWith({
				date: { $gte: new Date('2023-01-01'), $lte: new Date('2023-01-03') },
				status: TransactionStatus.COMPLETED,
			});
		});
	});

	describe('refreshTransactions', () => {
		it('should bulk upsert transactions', async () => {
			mockedMongoTransactionModel.bulkWrite.mockResolvedValue({ ok: 1 } as any);
			const transactions = multipleTransactionsFixture;

			const result = await repository.refreshTransactions(transactions);

			const bulkWriteArgs = mockedMongoTransactionModel.bulkWrite.mock.calls[0];
			expect(bulkWriteArgs).toMatchSnapshot();
			expect(result).toBe(true);
		});
	});
});
