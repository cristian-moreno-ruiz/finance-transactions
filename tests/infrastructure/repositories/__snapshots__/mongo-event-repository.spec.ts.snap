// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MongoTransactionRepository refreshTransactions should bulk upsert transactions 1`] = `
[
  [
    {
      "updateOne": {
        "filter": {
          "externalAccountId": 1234,
          "externalTransactionId": 291,
          "source": "SAMPLE_SOURCE",
        },
        "update": {
          "amount": -100,
          "currency": "EUR",
          "date": 2024-06-30T21:00:00.000Z,
          "description": "Payment for Rent",
          "externalAccountId": 1234,
          "externalTransactionId": 291,
          "id": "123e4567-e89b-12d3-a456-426614174000",
          "source": "SAMPLE_SOURCE",
          "status": "PENDING",
          "title": "Direct Debit",
        },
        "upsert": true,
      },
    },
    {
      "updateOne": {
        "filter": {
          "externalAccountId": 1234,
          "externalTransactionId": 1642,
          "source": "SAMPLE_SOURCE",
        },
        "update": {
          "amount": -50,
          "currency": "EUR",
          "date": 2024-10-01T20:00:00.000Z,
          "description": "Payment for Electricity",
          "externalAccountId": 1234,
          "externalTransactionId": 1642,
          "id": "123e4567-e89b-12d3-a456-426614174000",
          "source": "SAMPLE_SOURCE",
          "status": "PENDING",
          "title": "Electricity Bill",
        },
        "upsert": true,
      },
    },
  ],
]
`;
